types/karma/index.d.ts

```diff
@@ -301,25 +447,41 @@ export interface ConfigOptions {
      * [Express](http://expressjs.com/guide/using-middleware.html) docs).
      */
     middleware?: string[];
+    /**
+     * This is the same as middleware except that these middleware will be run before karma's own middleware.
+     * @default []
+     */
+    beforeMiddleware?: string[];
     /**
      * @default {}
      * @description Redefine default mapping from file extensions to MIME-type.
      * Set property name to required MIME, provide Array of extensions (without dots) as it's value.
      */
-    mime?: {[type: string]: string[]};
+    mime?: { [type: string]: string[] };
+    /**
+     * Socket.io pingTimeout in ms, https://socket.io/docs/server-api/#new-Server-httpServer-options.
+     * Very slow networks may need values up to 60000. Larger values delay discovery of deadlock in tests or browser crashes.
+     * @default 5000
+     */
+    pingTimeout?: number;
     /**
      * @default ['karma-*']
      * @description List of plugins to load. A plugin can be a string (in which case it will be required
      * by Karma) or an inlined plugin - Object.
      * By default, Karma loads all sibling NPM modules which have a name starting with karma-*.
      * Note: Just about all plugins in Karma require an additional library to be installed (via NPM).
      */
-    plugins?: any[];
+    plugins?: Array<string | { [plugin: string]: Array<string | ((...params: any[]) => any)> }>;
```

Comment:
Maybe change `[plugin: string]` to `[diToken: string]` or `[pluginName: string]` for clarity.

Also since this definition uses TypeScript 3.2 I think you can use [a tuple](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html#fixed-length-tuples) for inline plugin definition. I.e `[string, any]` as it should always accepts 2-element tuple. Also while docs only demonstrate `factory` provider it should potentially work with other [types of providers](https://github.com/vojtajina/node-di#registering-stuff).

Reply:
thaks, on this atm

Reply:
@devoto13 
This seems to work as you requested, mind commenting without me doing amend yet?
```diff
--- a/types/karma/index.d.ts
+++ b/types/karma/index.d.ts
@@ -215,6 +215,16 @@ export interface UpstreamProxy {
     protocol?: string;
 }
 
+// description of inline plugins
+export type PluginName = string;
+export type ConstructorFn = new (...params: any[]) => any;
+export type FactoryFn = (...params: any[]) => any;
+export type ConstructorFnType = ['constructor', ConstructorFn];
+export type FactoryFnType = ['factory', FactoryFn];
+export type ValueType = ['value', any];
+export type InlinePluginType = FactoryFnType | ConstructorFnType | ValueType;
+export type InlinePluginDef = Record<PluginName, InlinePluginType>;
+
 export interface ConfigOptions {
     /**
      * @description Enable or disable watching files and executing the tests whenever one of these files changes.
@@ -471,7 +481,7 @@ export interface ConfigOptions {
      * By default, Karma loads all sibling NPM modules which have a name starting with karma-*.
      * Note: Just about all plugins in Karma require an additional library to be installed (via NPM).
      */
-    plugins?: Array<string | { [plugin: string]: Array<string | ((...params: any[]) => any)> }>;
+    plugins?: Array<PluginName | InlinePluginDef>;
     /**
      * @default 9876
      * @description The port where the web server will be listening.
diff --git a/types/karma/karma-tests.ts b/types/karma/karma-tests.ts
index c9fe003fb3..45f8eed2fe 100644
--- a/types/karma/karma-tests.ts
+++ b/types/karma/karma-tests.ts
@@ -166,16 +166,17 @@ function CustomMiddlewareFactory(config: karma.ConfigOptions) {
     };
 }
 
-const factoryFn = () => {};
+class CustomPlugin {}
+
 const pluginsTests = (config: karma.Config) => {
     config.set({
         middleware: ['custom'],
         plugins: [
             'karma-jasmine',
             'karma-chrome-launcher',
-            { 'framework:xyz': ['factory', factoryFn] },
-            { 'middleware:custom': ['factory', CustomMiddlewareFactory] },
-            require('./plugin-required-from-config'),
+            { 'framework:xyz': ['factory', CustomMiddlewareFactory] },
+            { 'framework:abc': ['constructor', CustomPlugin] },
+            { 'framework:xyz': ['value', 5] },
         ],
     });
 };

```
Thanks!

Reply:
The only note is that `['constructor', CustomPlugin]` will fail at runtime. Because `di` module is very old, it does not support native class declarations. I would replace `class CustomPlugin {}` with `function CustomPlugin {}` to not confuse users. 

Otherwise looks very good 👍

Reply:
thanks, ok, revised one:
```diff
diff --git a/types/karma/index.d.ts b/types/karma/index.d.ts
index 6c51070980..008b52116d 100644
--- a/types/karma/index.d.ts
+++ b/types/karma/index.d.ts
@@ -215,6 +215,17 @@ export interface UpstreamProxy {
     protocol?: string;
 }
 
+// description of inline plugins
+export type PluginName = string;
+// tslint:disable-next-line:ban-types support for constructor function and classes
+export type ConstructorFn = Function | (new (...params: any[]) => any);
+export type FactoryFn = (...params: any[]) => any;
+export type ConstructorFnType = ['constructor', ConstructorFn];
+export type FactoryFnType = ['factory', FactoryFn];
+export type ValueType = ['value', any];
+export type InlinePluginType = FactoryFnType | ConstructorFnType | ValueType;
+export type InlinePluginDef = Record<PluginName, InlinePluginType>;
+
 export interface ConfigOptions {
     /**
      * @description Enable or disable watching files and executing the tests whenever one of these files changes.
@@ -471,7 +482,7 @@ export interface ConfigOptions {
      * By default, Karma loads all sibling NPM modules which have a name starting with karma-*.
      * Note: Just about all plugins in Karma require an additional library to be installed (via NPM).
      */
-    plugins?: Array<string | { [plugin: string]: Array<string | ((...params: any[]) => any)> }>;
+    plugins?: Array<PluginName | InlinePluginDef>;
     /**
      * @default 9876
      * @description The port where the web server will be listening.
diff --git a/types/karma/karma-tests.ts b/types/karma/karma-tests.ts
index c9fe003fb3..cec48db31c 100644
--- a/types/karma/karma-tests.ts
+++ b/types/karma/karma-tests.ts
@@ -166,16 +166,25 @@ function CustomMiddlewareFactory(config: karma.ConfigOptions) {
     };
 }
 
-const factoryFn = () => {};
+// plugin can be class or constructor function
+const CustomPlugin = function CustomPlugin() {};
+CustomPlugin.prototype = {
+    log: () => {},
+};
+class CustomPluginClass {
+    log: () => {};
+}
+
 const pluginsTests = (config: karma.Config) => {
     config.set({
         middleware: ['custom'],
         plugins: [
             'karma-jasmine',
             'karma-chrome-launcher',
-            { 'framework:xyz': ['factory', factoryFn] },
-            { 'middleware:custom': ['factory', CustomMiddlewareFactory] },
-            require('./plugin-required-from-config'),
+            { 'framework:xyz': ['factory', CustomMiddlewareFactory] },
+            { 'framework:abc': ['constructor', CustomPlugin] },
+            { 'framework:abc': ['constructor', CustomPluginClass] },
+            { 'framework:xyz': ['value', 5] },
         ],
     });
 };
```

Reply:
As I said, `{ 'framework:abc': ['constructor', CustomPluginClass] },` will not work with current version of Karma. But I'm fine keeping it in case it gets supported in the future.

Reply:
Also, just noticed, `'constructor'` should be `'type'` (see [docs](https://github.com/vojtajina/node-di#registering-stuff)).

Reply:
cool, nice catch, will amend now with that 'type' added, thanks!

Reply:
done, amended, thanks for looking into these details